extension/wallet/src/lib_wallet.rs

```diff
@@ -0,0 +1,348 @@
+    confidential,
+    secp256k1_zkp::SecretKey,
+    Address, AssetId, BlockHash, OutPoint, Transaction, TxOut, Txid,
+};
+use futures::{stream::FuturesUnordered, StreamExt, TryStreamExt};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+pub use crate::lib_wallet::{
+    extract_loan::{extract_loan, Error as ExtractLoanError},
+    extract_trade::extract_trade,
+    get_balances::get_balances,
+    get_transaction_history::get_transaction_history,
+    make_create_swap_payload::{
+        make_buy_create_swap_payload, make_sell_create_swap_payload, Error as MakePayloadError,
+    },
+    make_loan_request::{make_loan_request, Error as MakeLoanRequestError},
+    repay_loan::repay_loan,
+    sign_and_send_swap_transaction::sign_and_send_swap_transaction,
+    sign_loan::sign_loan,
+    withdraw_everything_to::withdraw_everything_to,
+};
+
+mod extract_loan;
+mod extract_trade;
+mod get_balances;
+mod get_transaction_history;
+mod make_create_swap_payload;
+mod make_loan_request;
+mod repay_loan;
+mod sign_and_send_swap_transaction;
+mod sign_loan;
+mod withdraw_everything_to;
+
+// TODO: make this configurable through extension option UI
+pub const DEFAULT_SAT_PER_VBYTE: u64 = 1;
+
+#[async_trait]
+pub trait Wallet {
+    async fn fetch_transaction(&self, txid: Txid) -> Result<Transaction>;
+    async fn fetch_transaction_history(&self, address: &Address) -> Result<Vec<Txid>>;
+    async fn get_fee_estimates(&self) -> Result<FeeEstimatesResponse>;
+    async fn broadcast(&self, tx: Transaction) -> Result<Txid>;
+    fn get_address(&self) -> Address;
+    fn blinding_key(&self) -> SecretKey;
+    fn secret_key(&self) -> SecretKey;
+    fn public_key(&self) -> PublicKey {
+        PublicKey::from_secret_key(SECP256K1, &self.secret_key())
+    }
+    /// A pure function to compute the balances of the wallet given a set of [`TxOut`]s.
+    fn compute_balances(&self, txouts: &[TxOut]) -> Vec<BalanceEntry> {
+        let grouped_txouts = txouts
+            .iter()
+            .filter_map(|utxo| match utxo {
+                TxOut {
+                    asset: confidential::Asset::Explicit(asset),
+                    value: confidential::Value::Explicit(value),
+                    ..
+                } => Some((*asset, *value)),
+                txout => match txout.unblind(SECP256K1, self.blinding_key()) {
+                    Ok(unblinded_txout) => Some((unblinded_txout.asset, unblinded_txout.value)),
+                    Err(e) => {
+                        log::warn!("failed to unblind txout: {}", e);
+                        None
+                    }
+                },
+            })
+            .into_group_map();
+
+        grouped_txouts
+            .into_iter()
+            .map(|(asset, utxos)| {
+                let total_sum = utxos.into_iter().sum();
+                BalanceEntry {
+                    asset,
+                    value: total_sum,
+                }
+            })
+            .collect()
+    }
```

Comment:
Why is this not just a stand-alone function? Seems unnecessary to depend on `self` if you could also just take in the blinding-key?